File: src/algorithms/dijkstra.py
import heapq
from .base import PathfindingAlgorithm


class Dijkstra(PathfindingAlgorithm):
    """Uniform-cost search: h(n) = 0."""

    @property
    def name(self) -> str:
        return "Dijkstra"

    @property
    def color(self) -> tuple:
        return (80, 120, 255)

    def find_path(self, grid, start: tuple, end: tuple) -> tuple:
        dist   = {start: 0}
        prev   = {start: None}
        closed = set()
        visited_order = []
        pq = [(0, start)]

        while pq:
            cost, node = heapq.heappop(pq)
            if node in closed:
                continue
            closed.add(node)
            visited_order.append(node)

            if node == end:
                break

            for neighbor, edge_cost in grid.get_neighbors_with_cost(node):
                new_cost = cost + edge_cost
                if neighbor not in dist or new_cost < dist[neighbor]:
                    dist[neighbor] = new_cost
                    prev[neighbor] = node
                    heapq.heappush(pq, (new_cost, neighbor))

        return _reconstruct(prev, start, end), visited_order, dist


def _reconstruct(prev: dict, start: tuple, end: tuple) -> list:
    path, node = [], end
    while node is not None:
        path.append(node)
        node = prev.get(node)
    path.reverse()
    return path if path and path[0] == start else []

File: web/src/algorithms/dijkstra.js
// Mirrors src/algorithms/dijkstra.py — uniform-cost search (h = 0).

import { PathfindingAlgorithm } from './base.js';
import { MinHeap } from '../utils.js';

export class Dijkstra extends PathfindingAlgorithm {
    get name()  { return 'Dijkstra'; }
    get color() { return [80, 120, 255]; }

    findPath(grid, start, end) {
        const dist   = new Map([[start, 0]]);
        const prev   = new Map([[start, null]]);
        const closed = new Set();
        const visitedOrder = [];
        const pq = new MinHeap();
        pq.push([0, start]);

        while (pq.size > 0) {
            const [cost, node] = pq.pop();
            if (closed.has(node)) continue;
            closed.add(node);
            visitedOrder.push(node);

            if (node === end) break;

            for (const [nbr, edgeCost] of grid.getNeighborsWithCost(node)) {
                const newCost = cost + edgeCost;
                if (!dist.has(nbr) || newCost < dist.get(nbr)) {
                    dist.set(nbr, newCost);
                    prev.set(nbr, node);
                    pq.push([newCost, nbr]);
                }
            }
        }

        return {
            path:         _reconstruct(prev, start, end),
            visitedOrder,
            costMap: dist,
        };
    }
}

// ── Shared path-reconstruction helper (also used by astar.js) ────────────────

export function reconstruct(prev, start, end) {
    return _reconstruct(prev, start, end);
}

function _reconstruct(prev, start, end) {
    const path = [];
    let node = end;
    while (node !== null && node !== undefined) {
        path.push(node);
        node = prev.get(node);
    }
    path.reverse();
    return (path.length > 0 && path[0] === start) ? path : [];
}
